crates/nu-cli/src/shell/filesystem_shell.rs

```diff
@@ -553,390 +553,51 @@ impl Shell for FilesystemShell {
     fn mv(
         &self,
         MoveArgs { src, dst }: MoveArgs,
-        name: Tag,
+        _name: Tag,
         path: &str,
     ) -> Result<OutputStream, ShellError> {
-        let name_tag = name;
-
         let path = Path::new(path);
         let source = normalize(path.join(&src.item));
-        let mut destination = normalize(path.join(&dst.item));
+        let destination = normalize(path.join(&dst.item));
 
-        let sources: Vec<_> = match glob::glob(&source.to_string_lossy()) {
-            Ok(files) => files.collect(),
-            Err(_) => {
-                return Err(ShellError::labeled_error(
-                    "Invalid pattern.",
-                    "invalid pattern",
-                    src.tag,
-                ))
-            }
-        };
+        let sources =
+            glob::glob(&source.to_string_lossy()).map_or_else(|_| Vec::new(), Iterator::collect);
 
         if sources.is_empty() {
             return Err(ShellError::labeled_error(
-                "Invalid file or pattern.",
+                "Invalid file or pattern",
                 "invalid file or pattern",
                 src.tag,
             ));
         }
 
-        let destination_file_name = {
-            match destination.file_name() {
-                Some(name) => PathBuf::from(name),
-                None => {
-                    let name_maybe =
-                        destination.components().next_back().and_then(
-                            |component| match component {
-                                Component::RootDir => Some(PathBuf::from("/")),
-                                Component::ParentDir => destination
-                                    .parent()
-                                    .and_then(|parent| parent.file_name())
-                                    .map(PathBuf::from),
-                                _ => None,
-                            },
-                        );
-
-                    if let Some(name) = name_maybe {
-                        name
-                    } else {
-                        return Err(ShellError::labeled_error(
-                            "Rename aborted. Not a valid destination",
-                            "not a valid destination",
-                            dst.tag,
-                        ));
-                    }
-                }
-            }
-        };
-
-        if sources.is_empty() {
+        // We have two possibilities.
+        //
+        // First, the destination exists.
+        //  - If a directory, move everything into that directory, otherwise
+        //  - if only a single source, overwrite the file, otherwise
+        //  - error.
+        //
+        // Second, the destination doesn't exist, so we can only rename a single source. Otherwise
+        // it's an error.
+
+        if (destination.exists() && !destination.is_dir() && sources.len() > 1)
+            || (!destination.exists() && sources.len() > 1)
```

Comment:
This is a pretty intense set of conditionals. I did this to do the early return, and only write the `for entry in sources` part once. Another option would be to move that part into a lambda/func, and write out the cases here as separate if/else if branches.